lang/src/org/partiql/lang/eval/EvaluatingCompiler.kt

```diff
@@ -437,17 +436,38 @@ internal class EvaluatingCompiler(
     }
 
     private fun compileNAryIn(
-        argThunks: List<ThunkEnv>,
+        args: List<ExprNode>,
         metas: MetaContainer): ThunkEnv {
-        val needleThunk = argThunks[0]
-        val haystackThunk = argThunks[1]
-        return thunkEnv(metas) { env ->
-            val needle = needleThunk(env)
-            val haystack = haystackThunk(env)
-            haystack.any { it.exprEquals(needle) }.exprValue()
+        val leftArg = compileExprNode(args[0])
+        val rightArg = args[1]
+        
+        return when {
+            // When the right arg is a list of literals we use a Set to speed up the predicate
+            rightArg is ListExprNode && rightArg.values.all { it is Literal } -> {
+                val inSet = rightArg.values
+                    .map { it as Literal }
+                    .fold(TreeSet<ExprValue>(DEFAULT_COMPARATOR)) { set, el ->
```

Comment:
Using a `fold` here seems unnecessarily complex, especially because `TreeSet<>` is a mutable data structure.

I would just declare `inSet` above as an empty `TreeSet<>` and then `.forEach` over `rightArg.values` to `.add` each value.

Reply:
forgot that kotlin has `.mapTo` will use that